### PR TITLE
Refactor bluetooth mapper to discover characteristics only once

### LINKED ISSUE #808: Bluetooth mapper should only discover the characteristics once
**What would you like to be added**:
Bluetooth mapper currently discovers characteristics each time before performing operations, this is not an efficient method. It should only discover it once at the beginning and reuse the same information 

**Why is this needed**:
Bluetooth mapper currently discovers characteristics each time before performing operations, this is not an efficient method and may cause performance issues.

### PR DESCRIPTION
**Pull request by sujithsimon22**
**What type of PR is this?**
 /kind cleanup

**What this PR does / why we need it**:
This PR contains code to ensure that the  bluetooth mapper discovers characteristics only once at the beginning and reuse the same information to perform actions

**Which issue(s) this PR fixes**:

Fixes #808

### CODE REVIEW

**Comment by sujithsimon22** [comment]
cc @sids-b @rohitsardesai83  for review

File: device/bluetooth_mapper/watcher/watcher.go [hunk 1]
```diff
@@ -100,8 +100,21 @@ func onPeripheralDisconnected(p gatt.Peripheral, err error) {
 //onPeripheralConnected contains the operations to be performed as soon as the peripheral device is connected
 func (w *Watcher) onPeripheralConnected(p gatt.Peripheral, err error) {
 	actionmanager.GattPeripheral = p
-	DeviceConnected <- true
 	helper.ChangeDeviceState("online", deviceID)
+	ss, err := p.DiscoverServices(nil)
+	if err != nil {
+		glog.Errorf("Failed to discover services, err: %s\n", err)
```
Comment (rohitsardesai83) [inline device/bluetooth_mapper/watcher/watcher.go]: we should return on error if we fail to discover services.

File: device/bluetooth_mapper/watcher/watcher.go [hunk 2]
```diff
@@ -100,8 +100,21 @@ func onPeripheralDisconnected(p gatt.Peripheral, err error) {
...
+	}
+	for _, s := range ss {
+		// Discovery characteristics
+		cs, err := p.DiscoverCharacteristics(nil, s)
+		if err != nil {
+			glog.Errorf("Failed to discover characteristics, err: %s\n", err)
```
Comment (rohitsardesai83) [inline device/bluetooth_mapper/watcher/watcher.go]: %s -> %v. also we can log the service name for which the characteristics could not be discovered.

Comment (sujithsimon22) [reply device/bluetooth_mapper/watcher/watcher.go]: Done, thanks

File: device/bluetooth_mapper/watcher/watcher.go (hunk 1 above)
Comment (sujithsimon22) [reply device/bluetooth_mapper/watcher/watcher.go]: Done, thanks

File: device/bluetooth_mapper/watcher/watcher.go [hunk 3]
```diff
@@ -100,8 +101,22 @@ func onPeripheralDisconnected(p gatt.Peripheral, err error) {
 //onPeripheralConnected contains the operations to be performed as soon as the peripheral device is connected
 func (w *Watcher) onPeripheralConnected(p gatt.Peripheral, err error) {
 	actionmanager.GattPeripheral = p
-	DeviceConnected <- true
 	helper.ChangeDeviceState("online", deviceID)
+	ss, err := p.DiscoverServices(nil)
+	if err != nil {
+		glog.Errorf("Failed to discover services, err: %s\n", err)
+		os.Exit(1)
```
Comment (rohitsardesai83) [inline device/bluetooth_mapper/watcher/watcher.go]: right now our mapper manages only one device at a time right ? if we support multiple devices , then the watcher should not exit if it fails to connect to one device.  are we tracking future mapper improvements via open issues ?
please create issues for all such improvements.

Comment (sujithsimon22) [reply device/bluetooth_mapper/watcher/watcher.go]: we have an open umbrella issue to track future improvements in the mapper. We can add support for multiple devices as one item in it

**Comment by sids-b** [comment]
@sujithsimon22 squash the commits please

**Comment by sujithsimon22** [comment]
> @sujithsimon22 squash the commits please

Done :)

**Pull request merged**